presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedInPredicateToJoin.java

```diff
@@ -0,0 +1,431 @@
+    {
+        FunctionCall countCall = new FunctionCall(
+                QualifiedName.of("count"),
+                Optional.<Window>empty(),
+                Optional.of(condition),
+                false,
+                ImmutableList.<Expression>of() /* arguments */
+        );
+        BigintType countResultType = BIGINT;
+        Symbol countSymbol = symbolAllocator.newSymbol(nameHint, countResultType);
+
+        return immutableEntry(
+                countSymbol,
+                new AggregationNode.Aggregation(
+                        countCall,
+                        new Signature("count", FunctionKind.AGGREGATE, countResultType.getTypeSignature()),
+                        Optional.<Symbol>empty() /* mask */
+                )
+        );
+    }
+
+    private static Expression isGreaterThan(Symbol symbol, long value)
+    {
+        return new ComparisonExpression(
+                ComparisonExpressionType.GREATER_THAN,
+                symbol.toSymbolReference(),
+                bigint(value)
+        );
+    }
+
+    private static Expression not(Expression booleanExpression)
+    {
+        return new NotExpression(booleanExpression);
+    }
+
+    private static Expression isNotNull(Symbol symbol)
+    {
+        return new IsNotNullPredicate(symbol.toSymbolReference());
+    }
+
+    private static Expression bigint(long value)
+    {
+        return new Cast(new LongLiteral(String.valueOf(value)), BIGINT.toString());
+    }
+
+    private static Expression booleanConstant(@Nullable Boolean value)
+    {
+        if (value == null) {
+            return new Cast(new NullLiteral(), BOOLEAN.toString());
+        }
+        return new BooleanLiteral(value.toString());
+    }
+
+    /**
+     * TODO consult comon parts with {@link com.facebook.presto.sql.planner.optimizations.TransformCorrelatedScalarAggregationToJoin.Rewriter#decorrelateFilters}
+     */
+    private static class DecorrelatingVisitor
+            extends PlanVisitor<PlanNode, Optional<Decorrelated>>
+    {
+        private final Lookup lookup;
+        private final Set<Symbol> correlation;
+
+        public DecorrelatingVisitor(Lookup lookup, Iterable<Symbol> correlation)
+        {
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            this.correlation = ImmutableSet.copyOf(requireNonNull(correlation, "correlation is null"));
+        }
+
+        private Optional<Decorrelated> visit(PlanNode reference)
+        {
+            checkArgument(reference instanceof GroupReference);
+            return lookup.resolve(reference).accept(this, reference);
+        }
+
+        @Override
+        public Optional<Decorrelated> visitProject(ProjectNode node, PlanNode reference)
+        {
+            if (isLocallyCorrelated(node)) {
+                // We are not able to decorrelate projection
```

Comment:
When I read that, I thought that this "not able" means "it is impossible". If it is possible (however hard) `// TODO` will be better.